File: examples/cpp/clustering.cpp
#include <torch/torch.h>
#include <butane/butane.h>

int main()
{
    torch::Tensor rdata = torch::empty({1e+6, 2}).uniform_(static_cast<double>(-1), static_cast<double>(1));
    rdata = rdata.to(torch::Device(torch::kCUDA));
    butane::clustring::MiniBatchKMeans kmeans(200, butane::KMeansPlusPlus, 1024, 500, 1e-4, 3);
    kmeans.fit(rdata);
    std::cout << kmeans.centroids() << std::endl;
    return 0;
}

File: examples/python/clustering.py
import torch
import butane
import matplotlib.pyplot as plt

import numpy as np

if __name__ == "__main__":

    rdata = torch.empty((int(1e+5), 2), device='cuda').uniform_(-1, 1)
    kmeans = butane.clustering.MiniBatchKMeans(n_centroids=200, init='kmeans++', max_iters=500, tol=1e-4, random_state=3)
    kmeans.fit(rdata)
    fig, ax = plt.subplots()
    centroids = kmeans.centroids.cpu().numpy()
    rdata = rdata.cpu().numpy()
    ax.scatter(centroids[:, 0], centroids[:, 1])
    ax.set_xlim(-1, 1)
    ax.set_ylim(-1, 1)
    plt.show()